### PR TITLE
feat(button-toggle): Add a focus overlay for button-toggle

### LINKED ISSUE #3014: Focus styling and hiding for mat-button-toggle
#### Bug, feature request, or proposal:
Implement focus indication for md-button-toggle, similar to md-button.

#### What is the expected behavior?
Several form elements such as md-button have a md-button-focus class bound to a function that detects whether the element has focus and it's appropriate to show the focus indicator. (This doesn't just style the :focus pseudoselector because an element can have focus even when the focus indicator should not be shown.) md-button-toggle should implement this similarly.

#### What is the current behavior?
md-button-toggle does not implement md-button-focus, nor does it have focus styling. I believe the current behavior is to not indicate focus for toggle buttons or button groups at all. Focus itself is handled correctly: exclusive button groups take focus, as do individual toggles alone and in multi-select groups.

#### What are the steps to reproduce?
Simple demo of the different toggle buttons, where you can test keyboard focus behavior alongside other widgets:
  http://plnkr.co/edit/uLky5EnpTukImH7yMZOO?p=preview

#### What is the use-case or motivation for changing an existing behavior?
I'm not sure if focus behaviors are specified by Material for toggle buttons. Without guidance to the contrary, I would assume that focus indication for toggles is appropriate and would be similar to md-buttons. A lack of focus indication is a poor experience.

#### Which versions of Angular, Material, OS, browsers are affected?
All browsers.

#### Is there anything else we should know?
Unlike action buttons, toggle buttons may need to indicate 1) toggle state, 2) focus, 3) active state, and 4) hover state, all with background color changes. My project has decided to represent toggle and focus using shading levels, where toggle+focus gets a double shade, just toggle and just focus are identical (for better and for worse), and active and hover override toggle/focus and use a different color. I don't know what'd be appropriate for the Material library default. I can't implement a custom focus style correctly without md-button-focus support. (The library could implement md-button-focus without taking a position on styling. :) )

### PR DESCRIPTION
**Pull request by tinayuangao**
Fixes #3014

### CODE REVIEW

File: src/lib/button-toggle/button-toggle.ts [hunk 1]
```diff
@@ -461,10 +468,31 @@ export class MdButtonToggle implements OnInit {
     event.stopPropagation();
   }
 
+  _onInputFocus() {
+    // Only show the focus / ripple indicator when the focus was not triggered by a mouse
```
Comment (mmalerba) [inline src/lib/button-toggle/button-toggle.ts]: maybe you could use the FocusOriginMonitor for this? https://github.com/angular/material2/blob/master/src/lib/core/style/focus-classes.ts

File: src/lib/button-toggle/button-toggle.scss [hunk 2]
```diff
@@ -47,3 +53,27 @@ $mat-button-toggle-border-radius: 2px !default;
...
+  top: 0;
+  left: 0;
+  bottom: 0;
+  right: 0;
+
+  @include cdk-high-contrast {
+    // Note that IE will render this in the same way, no
+    // matter whether the theme is light or dark. This helps
+    // with the readability of focused buttons.
+    background-color: rgba(white, 0.5);
```
Comment (mmalerba) [inline src/lib/button-toggle/button-toggle.scss]: should be part of the theme file

File: src/lib/button-toggle/button-toggle.scss [hunk 3]
```diff
@@ -47,3 +53,27 @@ $mat-button-toggle-border-radius: 2px !default;
...
   vertical-align: middle;
 }
+
+// Overlay to be used as a tint. Note that the same effect can be achieved by using a pseudo
+// element, however we use a proper DOM element in order to be able to disable the default
+// touch action. This makes the buttons more responsive on touch devices.
+.mat-button-toggle-focus-overlay {
+  // The button spec calls for focus on raised buttons (and FABs) to be indicated with a
+  // black, 12% opacity shade over the normal color (for both light and dark themes).
+  background-color: rgba(black, 0.12);
```
Comment (mmalerba) [inline src/lib/button-toggle/button-toggle.scss]: should this be in the theme file?

File: src/lib/button-toggle/button-toggle.html [hunk 4]
```diff
@@ -5,10 +5,14 @@
...
+         (focus)="_onInputFocus()"
          (change)="_onInputChange($event)"
          (click)="_onInputClick($event)">
 
   <div class="mat-button-toggle-label-content">
     <ng-content></ng-content>
   </div>
 </label>
+<!-- the touchstart handler prevents the overlay from capturing the initial tap on touch devices -->
+<div class="mat-button-toggle-focus-overlay" (touchstart)="$event.preventDefault()"></div>
```
Comment (mmalerba) [inline src/lib/button-toggle/button-toggle.html]: why do we want to disable the initial tap?

File: src/lib/button-toggle/button-toggle.ts (hunk 1 above)
Comment (tinayuangao) [inline src/lib/button-toggle/button-toggle.ts]: Thanks for the suggestion!

File: src/lib/button-toggle/button-toggle.scss (hunk 3 above)
Comment (tinayuangao) [inline src/lib/button-toggle/button-toggle.scss]: I think only theme related should be in theme file. Default is black which will not change with the theme.

Added theme color in the theme file.

File: src/lib/button-toggle/button-toggle.html (hunk 4 above)
Comment (tinayuangao) [inline src/lib/button-toggle/button-toggle.html]: This is the same as https://github.com/angular/material2/blob/master/src/lib/button/button.html

File: src/lib/button-toggle/button-toggle.scss (hunk 3 above)
Comment (mmalerba) [inline src/lib/button-toggle/button-toggle.scss]: Ah ok, I missed that these are the same for light & dark themes

Comment (kara) [inline src/lib/button-toggle/button-toggle.scss]: Can you replace this with a color from the _palettes file though? If one doesn't exist in $background, we can create one.

File: src/lib/button-toggle/button-toggle.scss [hunk 5]
```diff
@@ -47,3 +53,27 @@ $mat-button-toggle-border-radius: 2px !default;
...
+  border-radius: inherit;
+  pointer-events: none;
+  opacity: 0;
+  position: absolute;
+  top: 0;
+  left: 0;
+  bottom: 0;
+  right: 0;
```
Comment (kara) [inline src/lib/button-toggle/button-toggle.scss]: Use the `mat-fill` mixin here?

File: src/lib/button-toggle/button-toggle.spec.ts [hunk 6]
```diff
@@ -14,13 +14,15 @@ import {
...
 } from './button-toggle';
+import {FocusOriginMonitor} from '../core';
 
 
 describe('MdButtonToggle', () => {
 
   beforeEach(async(() => {
     TestBed.configureTestingModule({
       imports: [MdButtonToggleModule.forRoot(), FormsModule, ReactiveFormsModule],
+      providers: [FocusOriginMonitor],
```
Comment (kara) [inline src/lib/button-toggle/button-toggle.spec.ts]: Should this be added to the button toggle's module definition?

File: src/lib/button-toggle/_button-toggle-theme.scss [hunk 7]
```diff
@@ -1,13 +1,25 @@
 @import '../core/theming/palette';
 @import '../core/theming/theming';
 
+// Applies a focus style to an md-button-toggle element for each of the supported palettes.
+@mixin _mat-button-toggle-focus-color($theme) {
+  $foreground: map-get($theme, foreground);
+
+  .mat-button-toggle-focus-overlay {
+    background-color: mat-color($foreground, divider);
```
Comment (kara) [inline src/lib/button-toggle/_button-toggle-theme.scss]: Nit: Do you want to add another property as 'dark-overlay' or something? Using a property from "foreground" as a background is a bit confusing out of context.

**Pull request merged**